gnmi_subscribe/gnmi_subscribe.go

```diff
@@ -77,97 +79,160 @@ func main() {
+	encoding, err := parseEncoding(*encodingFormat)
+	if err != nil {
+		log.Exitf("Error parsing encoding: %v", err)
 	}
 
-	var subscriptionListMode gnmi.SubscriptionList_Mode
-	switch {
-	case *subscriptionPoll && *subscriptionOnce:
+	subscriptionListMode, err := subscriptionMode(*subscriptionPoll, *subscriptionOnce)
+	if err != nil {
 		flag.Usage()
-		log.Exitf("Only one of -once and -poll can be set")
-	case *subscriptionOnce:
-		subscriptionListMode = pb.SubscriptionList_ONCE
-	case *subscriptionPoll:
-		subscriptionListMode = pb.SubscriptionList_POLL
-	default:
-		subscriptionListMode = pb.SubscriptionList_STREAM
+		log.Exit(err)
 	}
 
-	var pbPathList []*pb.Path
-	for _, xPath := range xPathFlags {
-		pbPath, err := xpath.ToGNMIPath(xPath)
-		if err != nil {
-			log.Exitf("error in parsing xpath %q to gnmi path", xPath)
-		}
-		pbPathList = append(pbPathList, pbPath)
-	}
-	for _, textPbPath := range pbPathFlags {
-		var pbPath pb.Path
-		if err := proto.UnmarshalText(textPbPath, &pbPath); err != nil {
-			log.Exitf("error in unmarshaling %q to gnmi Path", textPbPath)
-		}
-		pbPathList = append(pbPathList, &pbPath)
+	pbPathList, err := parsePaths(xPathFlags, pbPathFlags)
+	if err != nil {
+		log.Exitf("Error parsing paths: %v", err)
 	}
-	subscriptions, err := assembleSubscriptions(pbPathList)
+
+	subscriptions, err := assembleSubscriptions(*streamOnChange, *sampleInterval, pbPathList)
 	if err != nil {
 		log.Exitf("Error assembling subscriptions: %v", err)
 	}
 
 	request := &pb.SubscribeRequest{
 		Request: &pb.SubscribeRequest_Subscribe{
 			Subscribe: &pb.SubscriptionList{
-				Encoding:     pb.Encoding(encoding),
+				Encoding:     encoding,
 				Mode:         subscriptionListMode,
 				Subscription: subscriptions,
+				UpdatesOnly:  *updatesOnly,
 			},
 		},
 	}
 	utils.PrintProto(request)
 	if err := subscribeClient.Send(request); err != nil {
 		log.Exitf("Failed to send request: %v", err)
 	}
+
 	switch subscriptionListMode {
 	case pb.SubscriptionList_STREAM:
-		stream()
+		if err := stream(subscribeClient); err != nil {
+			log.Exitf("Error using STREAM mode: %v", err)
+		}
 	case pb.SubscriptionList_POLL:
-		poll()
+		if err := poll(subscribeClient, *updatesOnly, pollUser); err != nil {
+			log.Exitf("Error using POLL mode: %v", err)
+		}
 	case pb.SubscriptionList_ONCE:
-		once()
+		if err := once(subscribeClient); err != nil {
+			log.Exitf("Error using ONCE mode: %v", err)
+		}
 	}
 }
 
-func stream() {
+func pollUser() {
+	fmt.Scanln()
+}
 
+func stream(subscribeClient gnmi.GNMI_SubscribeClient) error {
+	for {
+		res, err := subscribeClient.Recv()
+		if err == io.EOF {
+			log.Info("Stream closed by target")
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		switch res.Response.(type) {
+		case *pb.SubscribeResponse_SyncResponse:
+			log.Info("SyncResponse received")
+		case *pb.SubscribeResponse_Update:
+			utils.PrintProto(res)
+		default:
+			return errors.New("Unexpected response type")
```

Comment:
nit: no capitals at start of error string

Reply:
Sure! Changed